[herbi.sf8] HTTP 웹 서버 구현하기 2단계

herbi1411 opened the pull request:
안녕하세요!! WAS 미션 2단계 리뷰 요청드립니다.
과제 하면서 Spring의 편리함을 좀 더 깨닫게 된 것 같습니다!!
2단계 요구사항인 `로그인 기능, 동적 템플릿, Session` 기능을 구현했습니다.

- Thread Pool 적용 (50개)
    - FixedThreadPool을 통해 50개의 쓰레드를 만들어서 클라이언트의 요청을 처리했습니다.
- ConcurrentHashMap 사용
    - DB가 HashMap을 사용하는 것으로 단순화 된 상태인데, 동시성 이슈가 발생할 것 같아 ConcurrentHashMap을 사용하도록 변경했습니다.
- LomBok 사용
    - Builder, Getter 등을 편리하게 사용하기 위해 Lombok을 사용했습니다.
- DB를 Repository로 변경, 인터페이스와 구현체로 분리
    - 스프링 프로젝트 구조를 흉내내고자 DB 패키지를 repository로 변경했습니다.
- Custom Controller 만들어서 요청 1개당 메서드 1개 매핑
    - 커스텀 어노테이션을 만들어 http method, url을 통해  요청을 매핑했습니다.
    - 기존의 1단계에서는  enum을 사용해 매핑했는데 이러헥 하면 요청 하나당 클래스 1개를 만들어야 해서 좀더 가독성이 좋게 보이기 위해 변경했습니다.
- Not Found 구현
    - 요청한 URI에 대한 Request Handler가 없으면 404코드를 반환하도록 했습니다.
- Controller, Service, Repository가 global로 한번씩만 초기화 되도록 설정
    - 기존에는 요청에 따라 쓰레드가 생성되면 매번 클래스 인스턴스를 생성하는 방식이었습니다.
    - 이를 main 메서드에서 한번에 객체를 생성하고 쓰레드가 생성되면 run 메서드에 넘기는 방식으로 변경했습니다.
- Cookie 객체 생성
    - session 관리를 위해 cookie가 필요해서 객체로 생성하고 key-value를 가지는 map을 가지게 했습니다.

7주동안 업무로 바쁘신 와중에도 리뷰 남겨주셔서 정말 감사했습니다!!! 🙇‍♂️🙇‍♂️

herbi-sf8 committed:
refactor: Thread Pool 적용(50개 고정)

herbi-sf8 committed:
chore: log 오류 해결을 위해 SpringBoot 의존성 제거

- SpringBoot와 logback 라이브러리를 같이 사용하고 Console App을 실행하면 Log 구현체를 찾지 못하는 현상이 발생해서 SpringBoot 라이브러리 제거

herbi-sf8 committed:
refactor: HashMap -> ConcurrentHashMap 변경

- Multi Thread 환경에서 사용을 위해 ConcurrentHashMap으로 변경

herbi-sf8 committed:
style: Lombok 적용

herbi-sf8 committed:
refactor: Handler, Request의 문자열 상수처리, User에 Builder 패턴 적용

herbi-sf8 committed:
refactor: Response 문자열 상수처리

herbi-sf8 committed:
refactor: Request 매직 넘버 상수화

herbi-sf8 committed:
refactor: DB를 Repository로 변경, 인터페이스와 구현체 분리

herbi-sf8 committed:
refactor: Custom Controller 생성해서 request 처리, (DTO, Service) 계층 생성

herbi-sf8 committed:
feat: 로그인 기능

herbi-sf8 committed:
feat: 템플릿 엔진 적용

- 로그인 된 사용자가 user/list/ 로 접근하면 등록된 유저 목록 페이지가 동적으로 생성돼서 반환
- 로그인 되지 않았다면 login.html로 redirect

herbi-sf8 committed:
feat: Session 요구사항 구현

- SessionRepository가 Session을 map 형태로 가지고 있도록
- 한 Session에는 User를 포함해 여러 object 저장 가능
- 로그인한 유저가 로그인 요청을 하면 index로 Redirect

herbi-sf8 committed:
refactor: WebServer에서 HTTP 요청에 필요한 객체(Controller, Service, Repository) 생성하도록 변경

- 요청별로 생성되는 쓰레드에서 객체를 매번 생성하는 것이 아니라 한번만 생성하도록 수정

herbi-sf8 committed:
feat: 일치하는 Request Handler Method가 없으면 404 코드 반환

herbi-sf8 committed:
refactor: Request의 Cookie를 객체로 생성

- Cookie에 JSESSION 값 이외에 다른 값이 들어올 수 있으므로 Map으로 만들어서 관리

herbi-sf8 committed:
refactor: Request에서 사용되는 상수를 별도 파일로 분리

- 상수 개수가 많아서 가독성이 떨어지므로 상수 파일을 별도 생성

herbi-sf8 committed:
style: model 패키지 안에 있는 dto 클래스들을 dto 패키지 생성후 이동

herbi-sf8 committed:
test: Request 테스트 추가

- Request 객체 속성 값에 대한 테스트

herbi-sf8 committed:
docs: 2단계 내용 README.md 추가

maythree commented on src/main/java/dto/LoginRequest.java:
```diff
@@ -0,0 +1,24 @@
+    private final String password;
+
+    public static LoginRequest of(Map<String, String> queryString) {
+        return LoginRequest.builder()
```
userId와 password가 nullable이 아니라면 여기서 Null 체크 처리가 되어도 좀 더 좋을 것 같습니다.

herbi1411 replied:
요청 객체에서 null 체크를 해서 예외처리를 하면 더 좋을 것 같습니다. 인자로 받은 queryString에서 userid나 password 값이 없으면 예외를 던지도록 했습니다!

추가로 이런 경우 Bad Request를 던지면 더 좋을 것 같아서 일단 `IllegalArgumentException`을 던진 다음에 RequestHandler에서 Bad Request를 반환하도록 처리하도록 했습니다.

maythree replied:
수정하신 방식이라면 of 메서드를 사용할때만 null에 대한 체크가 되는 것인데, 의도하신 생각과 같으신걸까요??
결국 위의 수정방식은 LoginRequest의 userId와 password는 nullable이다는 그대로 유지됩니다.

herbi1411 replied:
지금 방식대로면 다른 메서드를 통해 Request를 생성할 때 Nullable이 돼서 의도하지 않은 버그가 발생할 것 같습니다!! 이 부분은 생각을 못했네요.

해결하고자 처음에는 Lombok의 `@NonNull`을 사용할까 생각했는데 그런 경우 null 값이 인자로 들어올 시 NPE가 발생해 400코드를 반환하기 쉽지 않을 것 같았습니다. 그래서 javax.validation의 `@NotNull`을 사용하려고 해봤는데 순수 java 앱에서 어노테이션만으로는 유효성 검사가 이루어지지 않는 것 같아서 (사용하는 쪽에서 validator를 사용해야 함) 이 방법으로도 해결하지 못했습니다.

결국 생성자에서 직접 Null 여부를 체크하는 방법으로 해결했습니다! 추가로 공백으로 입력이 들어올 때도 예외를 던지도록 했습니다. 이 방법도 필드가 추가 되면 생성자의 검증 부분을 수정해야하고 NotNull이라는 것을 생성자를 봐야 알 수 있다는 단점이 있는 것 같습니다..!

maythree commented on src/main/java/repository/SessionRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+import model.Session;
+
+public interface SessionRepository {
```
SessionRepository와 UserRepository 인터페이스를 구성하신 이유가 따로 있으실까요?

herbi1411 replied:
메모리로 DB 관리되는 부분을 나중에 실제 DB를 도입해서 해보려고 인터페이스를 만들었는데, 순수 JAVA로 DB 커넥션을 만들어서 하려면 처리할 부분이 많아질 것 같아 진행하지 못했습니다! 그래서 일단 인터페이스가 만들어져있지만 구현체가 1개밖에 없는 상황입니다. repository 사용할 때는 인터페이스에 의존하도록 했습니다.

maythree commented on src/main/java/webserver/WebServer.java:
```diff
@@ -1,33 +1,57 @@
-    private static final Logger logger = LoggerFactory.getLogger(WebServer.class);
     private static final int DEFAULT_PORT = 8080;
+    private static final int THREAD_NUMS = 50;
+    private static final ExecutorService threadPool = Executors.newFixedThreadPool(THREAD_NUMS);
```
newFixedThreadPool을 사용하신 이유가 따로 있으실까요??

herbi1411 replied:
주어진 코드대로 하면 요청마다 쓰레드를 생성하고 요청이 끝나면 쓰레드를 지우는 구조인데 비용이 클 것 같아서 톰캣 서버의 설정을 따라하고자 ThreadPool을 사용해봤습니다!

그런데 오늘 수업에서 ThreadPool 내용이 있었는데 가변적으로 Thread 개수를 조절하면 좋을 것 같아서 10개를 우선 만들고 200개까지 가변적으로 증가하도록 수정했습니다!! (브라우저로 페이지 옮겨가며 테스트 해봤는데 로컬환경에서 혼자 요청하다보니 10개에서 안늘어나는 것으로 확인했습니다.)

maythree replied:
가변적으로 수정하신 것은 잘하신 것 같습니다👏👏
그런데 제가 여기서 질문 드렸던 의도는 쓰레드풀 생성방법과 형식이 여러개가 있는데 관련해서 인지하고 사용을 하셨었는지가 궁금하여 질문을 드렸습니다. 시간 나실때 관련해서 학습을 추가로 진행해보시어도 좋으실 것 같습니다.
- newScheduledThreadPool, newFixedThreadPool, newCachedThreadPool, ForkJoinPool 등

herbi1411 replied:
쓰레드 풀을 관리하는데 더 다양한 기법들이 있군요. 학습해보겠습니다. 감사합니다!!😊

maythree reviewed (changes requested):
고생많으셨습니다.👍
앞으로도 지금처럼 프레임워크를 사용하실때 실제 구현은 어떻게 되는지에 대하여 이해를 해보시고 사용하신다면 성장하시는데 더욱 좋으실 것 같습니다.
그리고 패키지 구조가 일관되지 않는 것 같은데 이 부분은 다시 한번 고민해보시면 좋으실 것 같습니다.

herbi-sf8 committed:
refactor: LoginRequest 필드에 대한 Null 여부 체크 추가

herbi-sf8 committed:
refactor: 요청 처리 중 발생하는 에러 처리

- 잘못된 Request: 400
- 서버 내부 오류: 500
- Output 중 발생한 오류: 클라이언트에게 오류를 전달할 방법이 없어서 error log만 처리

herbi-sf8 committed:
refactor: RequestHandler 2중 depth 발생하는 부분 메서드로 분리

herbi-sf8 committed:
refactor: ThreadPool 쓰레드 개수 가변적으로 변하도록 수정

- core: 10개
- max: 200개
- 60초동안 운휴면 삭제

herbi-sf8 committed:
style: 패키지 구조 수정

maythree commented on src/main/java/dto/LoginRequest.java:
```diff
@@ -0,0 +1,24 @@
+
+    public static LoginRequest of(Map<String, String> queryString) {
+        return LoginRequest.builder()
+                .userId(queryString.get("userId"))
```
이거는 위에 상수 선언했으니 같이 바꿔주시는게....좋으실 것 같습니다😁

herbi1411 replied:
바꾸려고 위에 선언해놓고 아래에 안바꿨네요......... 감사합니다!

maythree commented on src/main/java/dto/UserRequest.java:
```diff
@@ -0,0 +1,35 @@
+    private final String name;
+    private final String email;
+
+    public static UserRequest of(Map<String, String> queryStringMap) {
```
이 부분도 LoginRequest와 동일하게 null에 대한 처리를 고민해보시면 좋으실 것 같네요.

herbi1411 replied:
넵 이부분도 생성자에서 유효성 검사를 하도록 처리했습니다!!😀

herbi-sf8 committed:
refactor: LoginRequest 생성자에서 인자값 Null 여부 체크

herbi-sf8 committed:
refactor: LoginRequest 문자열 상수 처리

herbi-sf8 committed:
refactor: UserRequest 문자열 상수처리, Null여부 체크

maythree reviewed (approved):
지금까지 과제하시니라 고생 많으셨습니다.👍
과제는 여기서 마무리하도록 하겠습니다. 나중에 시간 되실때 했던 것들에 대해서 회고도 한번 해보시면서 내재화를 좀 더 해보시면 좋으실 것 같습니다.
